File: app_codebase/dart_bkend_base_02/bin/core/state/room/broadcast_hub.dart
/// Tier 2 — broadcast server push to room members via [ConnectionRegistry].
library;

import '../../ws/response/ws_response.dart';
import '../connection_registry.dart';
import '../contracts/room_broadcaster_contract.dart';
import '../contracts/room_membership_contract.dart';

class BroadcastHub implements RoomBroadcasterContract {
  BroadcastHub({
    required ConnectionRegistry connections,
    required RoomMembershipContract membership,
  })  : _connections = connections,
        _membership = membership;

  final ConnectionRegistry _connections;
  final RoomMembershipContract _membership;

  @override
  void broadcastToRoom(
    String roomId, {
    required String channel,
    required String type,
    required Map<String, dynamic> payload,
    String? excludeConnectionId,
  }) {
    final frame = encodeWsOk({
      'type': type,
      'channel': channel,
      'payload': payload,
    });
    for (final connectionId in _membership.connectionIds(roomId)) {
      if (connectionId == excludeConnectionId) continue;
      _connections.send(connectionId, frame);
    }
  }
}

File: app_codebase/python_base_05/bin/core/state/room/broadcast_hub.py
"""Tier 2 — broadcast server push to room members."""

from __future__ import annotations

from core.state.connection_registry import ConnectionRegistry
from core.state.contracts.room_broadcaster_contract import RoomBroadcasterContract
from core.state.contracts.room_membership_contract import RoomMembershipContract
from core.ws.response.ws_response import encode_ok


class BroadcastHub(RoomBroadcasterContract):
    def __init__(
        self,
        *,
        connections: ConnectionRegistry,
        membership: RoomMembershipContract,
    ) -> None:
        self._connections = connections
        self._membership = membership

    def broadcast_to_room(
        self,
        room_id: str,
        *,
        channel: str,
        msg_type: str,
        payload: dict,
        exclude_connection_id: str | None = None,
    ) -> None:
        frame = encode_ok({"type": msg_type, "channel": channel, "payload": payload})
        for connection_id in self._membership.connection_ids(room_id):
            if connection_id == exclude_connection_id:
                continue
            self._connections.send(connection_id, frame)
